Sync/BatchingClient.swift

```diff
@@ -0,0 +1,245 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Alamofire
+import Shared
+import XCGLogger
+import Deferred
+
+// MARK: Batch Client Errors
+public class SerializeRecordFailure<T: CleartextPayloadJSON>: MaybeErrorType {
+    public let record: Record<T>
+
+    public var description: String {
+        return "Failed to serialize record: \(record)"
+    }
+
+    public init(record: Record<T>) {
+        self.record = record
+    }
+}
+
+private let log = Logger.syncLogger
+
+// MARK: Internal Types
+private typealias UploadRecord = (guid: GUID, payload: String, size: Int)
+private typealias DeferredResponse = Deferred<Maybe<StorageResponse<POSTResult>>>
+
+/**
+ *  A payload represents a block of records sent in a single POST to the server within a batch.
+ *  The payload tries to contain as many records it can fit while respecting the limits imposed
+ *  by the info/configuration endpoint.
+ */
+private struct Payload {
+    let config: InfoConfiguration
+
+    var isEmpty: Bool {
+        return records.isEmpty
+    }
+
+    private(set) var records: [UploadRecord] = []
+    private var bytes: ByteCount = 0
+    private var count: Int = 0
+
+    init(config: InfoConfiguration) {
+        self.config = config
+    }
+
+    mutating func add(record: UploadRecord) -> Bool {
+        guard (record.size + bytes + 1 <= config.maxPostBytes) && (count < config.maxPostRecords) else {
+            return false
+        }
+
+        records.append(record)
+        count += 1
+        bytes += record.size + 1     // For newline.
+        return true
+    }
+}
+
+private struct Batch {
+    let config: InfoConfiguration
+
+    // Batch is full if we can't add another payload to it
+    var isFull: Bool {
+        return (config.maxPostRecords + count > config.maxTotalRecords) ||
+               (config.maxPostBytes + bytes > config.maxTotalBytes)
```

Comment:
> 
>     The number of records we've already sent, plus another full POST, exceeds the total the server will accept, or
>     The number of bytes we've already sent, plus another full POST, exceeds the total the server will accept, then
>     We can't start another payload in this batch.

Correct.

> What if the next POST will only include one small record? We'll instead say this batch is full and commit it, moving on to another batch in the bundle, right?

Yup that's what would happen. I spoke with Grisha yesterday and he's doing the check at the record/byte level so we're not sending small payloads if they would have fit in the existing batch. This can definitely be made smarter.